####################################################################################
# patch to make output thumbnails use the frame number of the input
# file *instead* of an ever-incrementing rather meaningless counter
#
#   tracey jan2011 (last update: mar2012)
# 
####################################################################################
diff --git a/ffmpeg.c b/ffmpeg.c
index c48aec7..a6f92b6 100644
--- a/ffmpeg.c
+++ b/ffmpeg.c
@@ -829,6 +829,7 @@ static void do_video_out(AVFormatContext *s,
         av_log(NULL, AV_LOG_VERBOSE, "*** %d dup!\n", nb_frames - 1);
     }
 
+  s->ist_frame_number = ist->st->codec->frame_number;
   /* duplicates frame if needed */
   for (i = 0; i < nb_frames; i++) {
     av_init_packet(&pkt);
diff --git a/libavformat/avformat.h b/libavformat/avformat.h
index 166fdbb..2758269 100644
--- a/libavformat/avformat.h
+++ b/libavformat/avformat.h
@@ -1156,6 +1156,7 @@ typedef struct AVFormatContext {
      * to know how the duration was estimated.
      */
     enum AVDurationEstimationMethod duration_estimation_method;
+    int ist_frame_number;
 } AVFormatContext;
 
 /**
--- a/libavformat/img2enc.c
+++ b/libavformat/img2enc.c
@@ -66,7 +66,7 @@ static int write_packet(AVFormatContext *s, AVPacket *pkt)
 
     if (!img->is_pipe) {
         if (av_get_frame_filename(filename, sizeof(filename),
-                                  img->path, img->img_number) < 0 && img->img_number>1 && !img->updatefirst) {
+                                  img->path, (s->ist_frame_number ? s->ist_frame_number : img->img_number)) < 0 && img->img_number>1 && !img->updatefirst) {
             av_log(s, AV_LOG_ERROR,
                    "Could not get frame filename number %d from pattern '%s'\n",
                    img->img_number, img->path);
